core/src/upgrade/mod.rs

```diff
@@ -18,6 +18,45 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
+//! Contains everything related to upgrading a connection or a substream to use a protocol.
+//!
+//! After a connection with a remote has been successfully established or a substream successfully
+//! opened, the next step is to *upgrade* this connection or substream to use a protocol.
+//!
+//! This is where the `UpgradeInfo`, `InboundUpgrade` and `OutboundUpgrade` traits come into play.
+//! The `InboundUpgrade` and `OutboundUpgrade` traits are implemented on types that represent an
+//! collection of one or more possible protocols for respectively an ingoing or outgoing
+//! connection or substream.
+//!
+//! > **Note**: Multiple versions of the same protocol are handled the same way as multiple
+//! >           different protocols are. For example, `/foo/1.0.0` and `/foo/1.1.0` are treated as
+//! >           if they were totally unrelated.
+//!
+//! # Upgrade process
+//!
+//! An upgrade is performed in two steps:
+//!
+//! - A protocol negotiation step. The `UpgradeInfo::protocol_names` method is called to determine
+//!   which protocols are supported by the trait implementation. The `multistream-select` protocol
+//!   is used in order to agree on which protocol to use amongst the ones supported.
+//!
+//! - A handshake. After a successful negotiation, the `InboundUpgrade::upgrade` or
+//!   `OutboundUpgrade::upgrade` method is called. This method will return a `Future` that
+//!   performs a handshake. This handshake is considered mandatory, however in practice it is
+//!   possible for the trait implementation to return a dummy `Future` that doesn't perform any
+//!   action and immediately succeeds.
```

Comment:
q: what would a use-case for "skipping" the handshake like this be?

Reply:
Almost all protocols skip the handshake. The exceptions are secio and identify.
What I call the handshake is the step between the negotiation and the output.